applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <misc/byteorder.h>
+
+#define MODULE led_stream
+#include "module_state_event.h"
+
+#include <logging/log.h>
+
+#include "config_event.h"
+#include "led_event.h"
+
+#define STREAM_LED_ID 0
+#define QUEUE_SIZE 5
+#define STREAM_DATA_SIZE 7
+
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_LED_STREAM_LOG_LEVEL);
+
+static struct led_effect led_stream_effect = {
+	.step_count = 1,
+};
+static struct led_effect_step steps_queue[QUEUE_SIZE];
+static u8_t rx_idx;
+static u8_t tx_idx;
+static bool first_sent;
+static bool leds_initialized;
+
+static void queue_data(const struct event_dyndata *dyndata)
+{
+	LOG_INF("Insert_data_to_effect");
+	size_t pos = 0;
+
+	memcpy(steps_queue[rx_idx].color.c, &dyndata->data[pos],
+	       CONFIG_DESKTOP_LED_COLOR_COUNT);
+	pos += 3;
+	steps_queue[rx_idx].substep_count = sys_get_le16(&dyndata->data[pos]);
+	pos += sizeof(steps_queue[rx_idx].substep_count);
+	steps_queue[rx_idx].substep_time = sys_get_le16(&dyndata->data[pos]);
+}
+
+static void send_effect(struct led_effect *effect, size_t led_id)
+{
+	LOG_INF("Send_effect to leds");
+	struct led_event *led_event = new_led_event();
+
+	led_event->stream = true;
+	led_event->led_id = led_id;
+	led_event->led_effect = effect;
+
+	__ASSERT_NO_MSG(led_event->led_effect->steps);
+
+	EVENT_SUBMIT(led_event);
+}
+
+static u8_t count_free_places(void)
+{
+	if (rx_idx > tx_idx) {
+		return QUEUE_SIZE - (rx_idx - tx_idx);
+	} else if (rx_idx < tx_idx) {
+		return QUEUE_SIZE - (tx_idx - rx_idx);
+	} else {
+		return QUEUE_SIZE;
+	}
+}
+
+static void store_data(const struct event_dyndata *data)
+{
+	u8_t free_places = count_free_places();
+
+	if (free_places > 0) {
+
+		LOG_INF("Insert data on position %" PRIu8
+		", free places %" PRIu8, rx_idx, free_places);
+		queue_data(data);
+		rx_idx++;
+		if (rx_idx == QUEUE_SIZE) {
+			rx_idx = 0;
+		}
+	} else {
+		LOG_INF("Queue is full - dropping incoming step");
+	}
+}
+
+static void send_data_from_queue(void)
+{
+	u8_t free_places = count_free_places();
+
+	if (free_places < QUEUE_SIZE) {
+		LOG_INF("Sending data: ");
+		for (size_t i = 0; i < CONFIG_DESKTOP_LED_COLOR_COUNT; i++) {
```

Comment:
please remove this - it's good during the creation of the code but makes it harder to read and maintain